typescript/src/game_loop/game-rxjs.ts

```diff
@@ -1,115 +1,158 @@
+  if (!process.env.TICK_RATE) {
+    return 60;
+  }
+  return +process.env.TICK_RATE || 60;
+}
+export interface GameResults {
+  readonly stats: GameStat;
+  readonly winner: BaseSocket;
+  readonly loser: BaseSocket;
+  readonly disconnected: boolean;
 }
 
-export type GameResults = [GameStat, BaseSocket, BaseSocket];
-export function runRxJSLoop([s1, s2]: [RxSocket, RxSocket]): Observable<GameResults> {
-    return new Observable((observer: Subscriber<GameResults>) => {
-        const subs: Subscription[] = [];
-        const stats = new GameStat();
-        const queue = new GameQueueRxJSImpl(s1, s2);
-        const world = new GameWorld(s1, s2);
-        const loop = new GameLoopRxJS(getTickRate());
-
-        function close(other: RxSocket) {
-            if (world.done) {
-                return;
-            }
-
-            loop.stop();
-            world.stop();
-            other.push(errorGameOver("The other player disconnected"));
-            observer.error(new Error("Disconnected"));
-        }
+class GameResultsImpl implements GameResults {
+  constructor(
+    public readonly stats: GameStat,
+    public readonly winner: BaseSocket,
+    public readonly loser: BaseSocket,
+    public readonly disconnected: boolean
+  ) {}
+}
 
-        subs.push(s1.events.subscribe({
-            complete: () => {
-                close(s2);
-            }
-        }));
-
-        subs.push(s2.events.subscribe({
-            complete: () => {
-                close(s1);
-            }
-        }));
-
-        subs.push(loop.start().pipe(
-            tap((delta: number) => {
-                stats.addDelta(delta);
-
-                // 1. process messages
-                const msgs = queue.flush();
-                if (msgs) {
-                    msgs.forEach(m => world.processMessage(m.from, m.message));
-                }
-
-                // 2. update all positions
-                world.update(delta);
-
-                // 3. process collisions
-                world.collisions();
-
-            }),
-            tap(() => {
-                if (world.done) {
-                    loop.stop();
-                    world.stop();
-
-                    const gameResult: GameResults = [stats, world.getWinner(), world.getLoser()];
-                    observer.next(gameResult);
-                    observer.complete();
-                    subs.forEach(s => {
-                        s.unsubscribe();
-                    });
-                }
-            })).subscribe());
-    })
+export class MessageEnvelope {
+  constructor(
+    public readonly message: Message,
+    public readonly from: BaseSocket
+  ) {}
 }
 
-function empty<T>(): Observable<T> {
-    return new Observable((observer: Subscriber<T>) => {
-        observer.complete();
+export function runRxJSLoop(
+  subscriber: Subscriber<GameResults>,
+  s1: RxSocket,
+  s2: RxSocket
+): void {
+  const stats = new GameStat();
+  const queue: MessageEnvelope[] = [];
```

Comment:
I completely removed the `GameQueueRxJSImpl` type. All it was doing was pushing values into an array, then later copying the array out and resetting. The abstraction was unnecessary. Apples-to-apples, I'd recommend making this change in the callbacks version as well (which I didn't really look at, TBH)